internal/common/voting/VotingMethods.go

```diff
@@ -0,0 +1,304 @@
+
+	// ----- first round -----
+	// find the count number of each lootbox
+	for _, preference := range voteList {
+		var maxPreference float64
+		var firstLootBoxChoice uuid.UUID
+		for lootBox, key := range preference {
+			if key > maxPreference {
+				firstLootBoxChoice = lootBox
+			}
+		}
+		voteCount[firstLootBoxChoice]++
+	}
+
+	// find the two candidates with most first-placed votes
+	var maxVotes1, maxVotes2 float64
+	var winner1, winner2 uuid.UUID
+	for lootBox, votes := range voteCount {
+		if votes > maxVotes1 {
+			winner2 = winner1
+			maxVotes2 = maxVotes1
+			winner1 = lootBox
+			maxVotes1 = votes
+		} else if votes > maxVotes2 {
+			winner2 = lootBox
+			maxVotes2 = votes
+		}
+	}
+
+	// check if either already has a majority or we need the second round
+	if maxVotes1 >= (maxVotes2 * 2) {
+		// return the majority lootbox
+		return winner1
+	} else {
+		// ----- second round -----
+		voteCount := make(map[uuid.UUID]int)
+		for _, preference := range voteList {
+			if preference[winner1] > preference[winner2] {
+				voteCount[winner1]++
+			} else {
+				voteCount[winner2]++
+			}
+		}
+		if voteCount[winner1] > voteCount[winner2] {
+			winner = winner1
+		} else {
+			winner = winner2
+		}
+	}
+
+	return winner
+}
+
+func BordaCount(voteList []map[uuid.UUID]float64) uuid.UUID {
+	/*
+		BordaCount:
+			Each voter rank order all the candidates. With n candidates being ranked k scores (n-k)+1 Borda points.
+			The candidate with the highest Borda Score is the winner
+	*/
+	voteCount := make(map[uuid.UUID]float64)
+	var winner uuid.UUID
+
+	// initialise the map with all candidates
+	for _, preference := range voteList {
+		for key := range preference {
+			voteCount[key] = 0
+		}
+	}
+
+	// covert the unodered map into ordered list
+	var ss [][]kv
+	for _, preference := range voteList {
+		var s []kv
+		for k, v := range preference {
+			// ignore the lootbox if value is 0
+			if v != 0 {
+				s = append(s, kv{k, v})
+			}
+		}
+		// sort the list using preference value of each lootbox
+		sort.Slice(s, func(i, j int) bool {
+			// in the order from large to small
+			return s[i].Value > s[j].Value
+		})
+		ss = append(ss, s)
+	}
+
+	// calculate the Borda score for each candidates
+	for _, sortedList := range ss {
+		usedKeys := make(map[uuid.UUID]bool)
+		for i, kv := range sortedList {
+			score := float64(len(voteCount)) - float64(i) + 1
+			voteCount[kv.Key] += score
+			usedKeys[kv.Key] = true
+		}
+
+		// points shared if not explicity ranked
+		remainingKeyNumber := float64(len(voteCount)) - float64(len(sortedList))
+		remainingScore := (1 + remainingKeyNumber) * remainingKeyNumber / 2
+		for key := range voteCount {
+			if !usedKeys[key] {
+				voteCount[key] += remainingScore / remainingKeyNumber
+			}
+		}
+	}
+
+	// find the winner with highest score
+	var maxScore float64
+	for key, value := range voteCount {
+		if value > maxScore {
+			winner = key
+			maxScore = value
+		}
+	}
+
+	return winner
+}
+
```

Comment:
Same with this method - doesn't this require the agents to recursively revote? As in the vote is repeated again and again with one fewer agents to vote for each time.

Reply:
Revote is not needed for every method